cpp/src/arrow/dataset/dataset.cc

```diff
@@ -162,8 +157,21 @@ FragmentIterator InMemorySource::GetFragmentsImpl(
   return MakeMaybeMapIterator(std::move(create_fragment), get_batches_->Get());
 }
 
-FragmentIterator TreeSource::GetFragmentsImpl(std::shared_ptr<ScanOptions> options) {
-  return GetFragmentsFromSources(children_, options);
+Result<std::shared_ptr<UnionDataset>> UnionDataset::Make(std::shared_ptr<Schema> schema,
+                                                         DatasetVector children) {
+  for (const auto& child : children) {
```

Comment:
If I recall correctly this check is done at multiple places, so we should factor this out similarly like UnifySchemas

Reply:
I'm not sure what other checks you're referring to. If you link to them, I will extract a helper function

Reply:
E.g. https://github.com/apache/arrow/blob/master/cpp/src/arrow/table.cc#L478

I guess it will useful in the future.

Reply:
Doesn't need to implement in this PR, just have a jira before we forget it.

Reply:
Ah, I see. I thought you were referring to the dataset namespace specifically. I've added https://issues.apache.org/jira/browse/ARROW-7946 to address this